src/dns.rs

```diff
@@ -0,0 +1,464 @@
+use crate::topology::config::GlobalOptions;
+use futures::{Async, Future, Poll};
+use hyper::client::connect::dns::{Name, Resolve};
+use std::io;
+use std::net::ToSocketAddrs;
+use std::net::*;
+use std::vec::IntoIter;
+use trust_dns_resolver::config::*;
+use trust_dns_resolver::error::ResolveErrorKind;
+use trust_dns_resolver::AsyncResolver;
+use trust_dns_resolver::BackgroundLookupIp;
+
+/// Default port for DNS service.
+const DNS_PORT: u16 = 53;
+
+/// Default version will behave identically to std::net::ToSocketAddrs implementations.
+#[derive(Clone, Default)]
+pub struct DnsResolver(Option<AsyncResolver>);
+
+impl DnsResolver {
+    /// (Self,errors)
+    pub fn new(config: &GlobalOptions) -> Result<Self, String> {
+        if config.dns.is_empty() {
+            return Ok(DnsResolver::default());
+        }
+
+        let mut resolve_config = ResolverConfig::new();
+
+        let mut errors = Vec::new();
+
+        for s in config.dns.iter() {
+            let parsed = s
+                .parse::<SocketAddr>()
+                .or_else(|_| s.parse::<IpAddr>().map(|ip| SocketAddr::new(ip, DNS_PORT)));
+            match parsed {
+                Ok(socket_addr) => resolve_config.add_name_server(NameServerConfig {
+                    socket_addr,
+                    protocol: Protocol::Udp,
+                    tls_dns_name: None,
+                }),
+                Err(error) => errors.push(format!(
+                    "DNS server IP {:?} is invalid. Parse Error: {:?}",
+                    s, error
+                )),
+            }
+        }
+
+        if !errors.is_empty() {
+            return Err(format!(
+                "DNS resolver initialization failed with errors: {:?}",
+                errors
+            ));
+        }
+
+        let mut options = ResolverOpts::default();
+        options.attempts = 2;
+        options.validate = false;
+        options.ip_strategy = LookupIpStrategy::Ipv4AndIpv6;
+        options.use_hosts_file = false;
+        options.distrust_nx_responses = false;
+        options.num_concurrent_reqs = 1;
```

Comment:
Why 1 and say not 4?

Reply:
There is a bug in dns library. I am not sure of the cause, but with multiple concurrent requests some DNS servers aren't used.

Still, multiple name resolutions are done in parallel.

Reply:
I'm not aware of a bug in this area, could you clarify?

Reply:
In  `fn multiple_dns_servers` test, with 2 concurrent requests, 3 DNS servers, and three name resolutions done one after other, sometimes first request fails, sometimes second one, sometimes third one, and other times they all pass. And with 1 concurrent request, they all pass.

All three DNS servers serve different domains, and all three requests resolve different names.

Reply:
Have you tested this with distrust-nx-domain by any chance?

Reply:
Yes, it still behaves the same.

Reply:
@ktff can you open an issue in `trust-dns` for this bug if there isn't one? And we should make sure we reference back here.

Reply:
Will do